src/main/java/tigase/pubsub/repository/PresenceCollectorRepository.java

```diff
@@ -26,129 +27,179 @@
+		return entriesByService.values();
+	}
+
+	public Stream<UserResourceEntry> userResourceEntryStream() {
+		return entriesByService.values()
+				.stream()
+				.flatMap(ServiceEntry::userEntriesStream)
+				.flatMap(UserEntry::userResourceEntriesStream);
+	}
+
+	public Stream<UserResourceEntry> expiredUserResourceEntriesStream(long expirationTimestamp) {
+		return userResourceEntryStream().filter(
+				userResourceEntry -> userResourceEntry.isOlderThan(expirationTimestamp));
+	}
+
+	public class ServiceEntry {
+		private final BareJID serviceJid;
+		private final ConcurrentHashMap<BareJID, UserEntry> usersEntries = new ConcurrentHashMap<>();
+
+		public ServiceEntry(BareJID serviceJid) {
+			this.serviceJid = serviceJid;
+		}
+
+		public String add(JID jid, String caps) {
+			return synchronizeOnUserJID(jid.getBareJID(), () -> usersEntries.computeIfAbsent(jid.getBareJID(), k -> new UserEntry(serviceJid, k))
+					.add(jid.getResource(), caps));
+		}
+
+		public boolean remove(JID jid) {
+			return synchronizeOnUserJID(jid.getBareJID(), () -> {
+				UserEntry entries = usersEntries.get(jid.getBareJID());
+				if (entries == null) {
+					return false;
+				}
+				boolean result = entries.remove(jid.getResource());
+				if (entries.isEmpty()) {
+					usersEntries.remove(jid.getBareJID());
+				}
+				return result;
+			});
+		}
+
+		public UserEntry get(BareJID jid) {
+			return usersEntries.get(jid);
+		}
+
+		public Collection<UserEntry> getUserEntries() {
+			return usersEntries.values();
+		}
 
-		if (resources != null) {
-			return resources.remove(resource);
+		public Stream<UserEntry> userEntriesStream() {
+			return usersEntries.values().stream();
 		}
 
-		return false;
+		protected <T> T synchronizeOnUserJID(BareJID jid, Supplier<T> run) {
+			synchronized (JID_LOCKS[Math.abs(jid.hashCode()) % JID_LOCKS.length]) {
+				return run.get();
+			}
+		}
 	}
 
-	public class Entries {
+	public class UserEntry {
 
-		private ArrayList<Entry> entries = new ArrayList<>();
+		private final BareJID serviceJid;
+		private final BareJID jid;
+		private final CopyOnWriteArrayList<UserResourceEntry> entries = new CopyOnWriteArrayList<>();
 
-		public Entries()  {
+		public UserEntry(BareJID serviceJid, BareJID jid)  {
+			this.serviceJid = serviceJid;
+			this.jid = jid;
+		}
+
+		public BareJID getJid() {
+			return jid;
+		}
 
+		public BareJID getServiceJid() {
+			return serviceJid;
 		}
 
-		public synchronized String[] add(String resource, String[] caps) {
-			String[] oldCaps = null;
+		public synchronized String add(String resource, String caps) {
+			String oldCaps = null;
 			for (int i=0; i<entries.size(); i++) {
-				Entry e = entries.get(i);
+				UserResourceEntry e = entries.get(i);
 				if (e.matches(resource)) {
 					oldCaps = entries.remove(i).caps;
 					break;
 				}
 			}
-			entries.add(new Entry(resource, caps));
+			// limit number of kept last available resources
+			while (entries.size() >= maximalNoOfResources) {
```

Comment:
Maybe we could re-utilize `CircularFifoQueue` here? Though looking at that class now it could have some issues and probably would need some improvements on it's own…
Otherwise should be ok.

Reply:
I was considering the usage of queue and list and I think `List` was simplest. Moreover, maybe we should change `List` to `LinkedMap` to remove "oldest", keep them sorted and "faster" on lookup then `List`.

Reply:
+1

that would definitely be the most sensible option as it handles capacity limit on it's own and helps addition/removel.